### PR TITLE
Migrate to jest 2

### PR DESCRIPTION
**Pull request by AtofStryker**
A rebase of #1587 with browser tests. 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [x] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue-test-utils/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

This branch introduces moving VTU to jest over mocha, as well as using a few different strategies to run in the browser (discussed below)

### CODE REVIEW

**Comment by AtofStryker** [comment]
@lmiller1990 I took a stab at getting this to work as mentioned in [1628](https://github.com/vuejs/vue-test-utils/pull/1628#issuecomment-667527786). In short, I took the changes from #1587 and rebased them on #1628 / dev. Overall, a pretty smooth rebase. 

It looks like the [tests](https://app.circleci.com/pipelines/github/vuejs/vue-test-utils/234/workflows/101e413b-7a36-4d3d-ba43-301505e465d9/jobs/14806) were failing to a `Error: write EPIPE` exception, which is almost impossible to diagnose. I found some evidence to point to issues with babel, in particular `@babel/code-frame` not supporting duplicate instances/packages. On a hunch, I [deduped](https://github.com/vuejs/vue-test-utils/pull/1629/commits/4d528031c221120c4bf29bd60fdd478eae0f153a#diff-8ee2343978836a779dc9f8d6b794c3b2R11) the `yarn.lock` file with [`yarn-deduplicate`](https://www.npmjs.com/package/yarn-deduplicate) and it seemed to fix the issue. :tada: 

I added a few commits that remove some of the old browser testing infrastructure as [other options](https://github.com/vuejs/vue-test-utils/pull/1587#issuecomment-647106762) are explored, which should allow for a fresh perspective. I might take a look into this to see what the jest + puppeteer solution looks like that you mentioned :smile: . If this isn't desired, the commits can be rebased out.

While running this in CI, I ran into `ENOMEM: not enough memory, read` from jest. I tried increasing the memory of the node instance, but the container only goes up to 4GBs, which still wasn't enough. I opted for jest to use 1 worker in CI in 4d52803 since it shouldn't slow the tests down that much, especially considering the jest cache.

Other commits are mainly doing some shuffling to represent what was present in #1587 without compromising the rebase. I also re added the `vuepress` npm scripts to get the netlify jobs to pass.

Not sure if you want to update #1587 with this branch. I only created this draft since I am working through a forked version of the repo, and this was the only way I could think of to run the circle tests :rofl:. BTW great work on the jest upgrade!

**Comment by lmiller1990** [comment]
Wow, thanks a lot! I was stuck on `Error: write EPIPE`... as you said, no clear way to diagnose.

will close my branch in favor of this one. Before we merge this we still have some work to do:

- update to latest rollup (rollup 2, I guess?)
- we may want to consider some way of running the tests in a browser (using puppeteer?) (previous was using karma, we could consider that too, I don't have a strong opinion. I have had problems running puppeteer on Circle CI before).
- that said I don't think running the tests in a browser is really a blocker.
- Assuming netlify auto deploy for docs will continue working fine, but will need to verify this. 

Since you got Jest passing, that was the main blocker. I will work on the rollup migration in the next few days.

I think we can do a release before this, though, for the `setProps` bugfix you made. I think we should approach this migration  with caution, since it's quite a lot of under the hood changes.

**Comment by AtofStryker** [comment]
> Wow, thanks a lot! I was stuck on Error: write EPIPE... as you said, no clear way to diagnose. will close my branch in favor of this one. 

@lmiller1990  glad we figured it out and always happy to help! I really appreciate the inclusion and friendliness you have provided. I'll can update the description with some of the items you had in the previous PR to provide a bit more context
 
> Before we merge this we still have some work to do: update to latest rollup (rollup 2, I guess?)

Definitely a good idea!  Is that something we need to handle in this PR, or does it make sense to handle in another PR? Either way, sounds like it needs to be done.

>  we may want to consider some way of running the tests in a browser (using puppeteer?) (previous was using karma, we could consider that too, I don't have a strong opinion. I have had problems running puppeteer on Circle CI before).
> that said I don't think running the tests in a browser is really a blocker.

I explored a few options, especially `puppeteer`. The biggest issue I ran into with [`puppeteer`](https://github.com/puppeteer/puppeteer) and [`jest-puppeteer`](https://github.com/smooth-code/jest-puppeteer) is that it doesn't run in the browser, but proxies all of its commands to the browser through the devtools protocol. That means all the tests would need to execute in some function, such as [`page.evaluate`](https://github.com/puppeteer/puppeteer/blob/v5.2.1/docs/api.md#pageevaluatepagefunction-args). Looks like someone has [attempted](https://stackoverflow.com/questions/52892441/run-jest-unit-tests-in-real-browser-environment-with-window-access-using-puppete) this before, but didn't have luck. I also tried it and was spinning my wheels.

[Cypress](https://www.cypress.io/) may also be an option, but I didn't pursue it too strongly. I love Cypress, but I don't think it's applicable to the current use case.

Instead, I pursued an option laid out in this awesome [blog post](https://github.com/tom-sherman/blog/blob/master/posts/02-running-jest-tests-in-a-browser.md#the-good-stuff) while reading through this jest [issue]( https://github.com/facebook/jest/issues/139#issuecomment-665086682). Essentially, this allows us to use Jasmine as the test runner, since it's nearly identical to jest's test runner, while leveraging Jest's expect/assertion and mock/stubbing libraries in the browser. It still uses karma to build everything and run the test, but considering the options, this looks best for the use case and gets the job done.

I added certain setup files back again as well as npm scripts, but with the dependencies being the current major release and updating the configurations accordingly (webpack 4, loaders, etc). 

Instead of PhantomJS, this is using headless chrome in CI. Locally, chrome can be used and the tests can be seen executing to some degree. For CI, I installed puppeteer to gain the chromium installs so we can run inside a container.

`test-compat` is working fine for `Vue` `2.2` - `2.6`. I had a few compilation issues with the updated dependencies/configurations for older versions, so I needed to skip some tests in those conditions. For `2.0` and `2.1`, I could not get webpack to compile. I am guessing this had to do with deprecated support from vue loader or another dependency, but am not entirely sure on that. Currently `2.0` and `2.1` are not being run in CI, which I believed to be a fair trade off, considering we get better test coverage for various vue versions in different scenarios.

I took a stab at this in df19c1b and would love it if you could take a look!

> Since you got Jest passing, that was the main blocker. I will work on the rollup migration in the next few days.

Sweet! I will probably make a few eager PRs, like upgrading `vuepress` for the docs since the older versions seem to have a lot of dependency clashes. This is also probably a good time to verify the netlify deploys.

> I think we can do a release before this, though, for the `setProps` bugfix you made. I think we should approach this migration with caution, since it's quite a lot of under the hood changes.

:100: . Definitely want to take our time and make sure we do it right. I don't see why we couldn't do a release beforehand :smile: .

**Comment by AtofStryker** [comment]
I can also throw df19c1b in a separate PR since there is already a lot going on here. It's just a bit difficult to do with a fork but I can easily do it as a contributor. Would it be OK to add me as a contributor?

**Comment by lmiller1990** [comment]
Wow, incredible work! This would have taken me much longer. 

I do not have the power to add contributors unfortunately. Can we just add the content of df19c1b and get this merged?

So just to sum up:

- we now have tests running in a browser via Jest + Jasmine combo (awesome).
- there are problems with Vue 2.0 and 2.1 tests (you can still use the lib with 2.0 and 2.1, of course). I agree this trade off it worth it. 
- I had a look at df19c1b. Reviewing configuration is very difficult but if it's all passing than it should be ✅ great job!! merge !== release, and since the actual build process (rollup) did not change there should be little risk.

I guess my only question is is there anything blocking merging this right now? It sounds like you have finished the job - while upgrading rollup would be nice, it shouldn't block this PR, right?

**Comment by AtofStryker** [comment]
> Wow, incredible work! This would have taken me much longer.

@lmiller1990 Thanks! COVID-19 in the USA has left me with a bit more time than I would usually have available :sweat_smile: 

> I do not have the power to add contributors unfortunately. Can we just add the content of [df19c1b](https://github.com/vuejs/vue-test-utils/commit/df19c1b637c9b1e8c6a0e5259851412ae8745b22) and get this merged?

All good! I figured it couldn't hurt to ask :smile:. We could merge df19c1b here as is as long as you and the team are comfortable with it.

> So just to sum up:
> 
> * we now have tests running in a browser via Jest + Jasmine combo (awesome).

:heavy_check_mark: 
> * there are problems with Vue 2.0 and 2.1 tests (you can still use the lib with 2.0 and 2.1, of course). I agree this trade off it worth it.

:heavy_check_mark: 
> * I had a look at [df19c1b](https://github.com/vuejs/vue-test-utils/commit/df19c1b637c9b1e8c6a0e5259851412ae8745b22). Reviewing configuration is very difficult but if it's all passing than it should be white_check_mark great job!! merge !== release, and since the actual build process (rollup) did not change there should be little risk.

:heavy_check_mark: 
Summary seems spot on IMO! There is definitely a lot going on from a configuration standpoint, but the good news is it is pretty similar to what was previously standing. If there is something I can do, such as documentation or other additions, to make the configuration a bit more review-able/digestible, don't hesitate to let me know! Since we are trying to make this improvement last as long as possible (at least for VTU 1.x), we definitely want to make sure other contributors can consume/digest it. 

There shouldn't be any immediate risk to these changes, as the bundles are the exact same between branches. I would double check me on that though :wink: . I am in no rush to release or merge by any means. Whatever it takes for us to ship this with confidence and to do it right.

> I guess my only question is is there anything blocking merging this right now? It sounds like you have finished the job - while upgrading rollup would be nice, it shouldn't block this PR, right?

The only thing that I believe is of immediate concern is the removed build/release scripts in https://github.com/vuejs/vue-test-utils/pull/1629/commits/28fffd83e5f517cb8bf500bcd65d5c3f7c053eec#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L9 which can be seen  [here](https://github.com/vuejs/vue-test-utils/pull/1629/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L21) in the file diff. We can add that back in pretty easily, like [here](https://github.com/AtofStryker/vue-test-utils/commit/ec669ae9226969b73426babff83d52c735271241) before merging.

**Comment by AtofStryker** [comment]
As far as the rollup upgrade is concerned, it isn't necessary for this release, which we could merge it eagerly after just to make it easier to review separate from this chunk of changes. I had some time to look into the upgrade. The good news for us is that virtually nothing has changed for us between the plugin updates and [rollup 2.x](https://github.com/rollup/rollup/releases/tag/v2.0.0)!

I performed those upgraded changes in https://github.com/AtofStryker/vue-test-utils/commit/57c4a9c536a1fe224e2d7f66cf9ce9eb4a01af55 (Sorry, I couldn't help myself :rofl: ) and made a [gist](https://gist.github.com/AtofStryker/fbf3855b8ab43a1887aae004f0fed247#file-server-test-utils-diff) of that diff for the team/community review. None of the `semver` or `acorn` variables are actually used in the current bundle, and the other changes can be explained by https://github.com/rollup/plugins/blame/master/packages/commonjs/src/helpers.js#L38 and https://github.com/rollup/plugins/blob/master/packages/commonjs/src/helpers.js#L58 / https://github.com/rollup/plugins/pull/206. That being said, when we do perform this upgrade, we should test the build and check for regressions, and I would recommend a minor release just in the rare case we would need to patch.

**Comment by AtofStryker** [comment]
I also have a few other upgrades we can incorporate in separate PRs, including [vuepress](https://github.com/AtofStryker/vue-test-utils/commit/d18c4582069bf4bddac9fafee4ec620b79ad1328), [core-js 3](https://github.com/AtofStryker/vue-test-utils/commit/08caa8c242fecf40ef8e24f7009389dfa9c3429d) (previously blocked by old vuepress and corejs 2.x is deprecated), and [chalk](https://github.com/AtofStryker/vue-test-utils/commit/30b010e75fe0dbfc779614346991977b748a7deb) (that was definitely the easiest one!). I actually don't think we need `jsdom-global` or `jsdom` any longer in the workspace as it ships with jest and was removed [here](https://github.com/vuejs/vue-test-utils/pull/1629/files#diff-8a37b02ec23196b56ac7855273a89fd0L4). I might work on an eslint upgrade, and that should really be the final thing we need to do to get VTU 1.x dependencies in a more modern state (for now :speak_no_evil: )!

**Comment by AtofStryker** [comment]
> The only thing that I believe is of immediate concern is the removed build/release scripts in 28fffd8#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L9 which can be seen here in the file diff. We can add that back in pretty easily, like here before merging.

I can add that commit to this branch, update the description, and mark it ready for review if that is the course of action we want to take :smile:

**Comment by lmiller1990** [comment]
@AtofStryker that'd be great, thanks a lot! Thank you very much for "owning" this feature 👍 excited to drop a tweet about this when we merge it, lmk your Twitter if you have one so I can tag you.

File: packages/test-utils/src/wrapper.js [hunk 1]
```diff
@@ -716,7 +716,9 @@ export default class Wrapper implements BaseWrapper {
           const isUpdated = Object.keys(data).some(key => {
             return (
               // $FlowIgnore : Problem with possibly null this.vm
-              this.vm[key] === data[key] || this.vm.$attrs[key] === data[key]
+              this.vm[key] === data[key] ||
+              // $FlowIgnore : Problem with possibly null this.vm
+              (this.vm.$attrs && this.vm.$attrs[key] === data[key])
```
Comment (AtofStryker) [inline packages/test-utils/src/wrapper.js:721]: I had a few issues in CI recently in regards to this guy in vue 2.3. Looks like I introduced a small regression in VTU for `setProps` for vue version 2.3 in #1618. The stack trace was a bit difficult to track down, but this should fix the issues seen

**Comment by AtofStryker** [comment]
> @AtofStryker that'd be great, thanks a lot! Thank you very much for "owning" this feature +1 excited to drop a tweet about this when we merge it, lmk your Twitter if you have one so I can tag you.

@lmiller1990 sure thing! I appreciate you letting me be so involved in this and being so welcome to contributions. I'm a bad technologist and rarely touch social media these days :laughing: . But I would definitely appreciate the tag. My twitter is https://twitter.com/AtofStryker.  

As far as the PR goes. I made a few changes to address a regression [here](https://github.com/vuejs/vue-test-utils/pull/1629#discussion_r467534095), but [exercised CI](https://app.circleci.com/pipelines/github/vuejs/vue-test-utils?branch=pull%2F1629) after changes and it looks good to me. Build scripts are re added and I removed the `jsdom` and `jsdom-global` dependencies as they are no longer used. I think we are ready to rock and roll :guitar:

**Comment by lmiller1990** [comment]
I will get this merged this week.

**Comment by lmiller1990** [comment]
Ok! It seems like everything is working.

I am having many problems running this locally though. Linting failing...

```
$ eslint --ext js,vue .
Cannot read property 'range' of null
TypeError: Cannot read property 'range' of null
    at SourceCode.getTokenBefore (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/token-store/index.js:303:18)
    at checkSpacingBefore (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/rules/template-curly-spacing.js:52:42)
    at TemplateElement (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/rules/template-curly-spacing.js:117:17)
    at /Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/util/safe-emitter.js:47:58
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/util/safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/util/node-event-generator.js:251:26)
    at NodeEventGenerator.applySelectors (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/util/node-event-generator.js:280:22)
    at NodeEventGenerator.enterNode (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/util/node-event-generator.js:294:14)
    at CodePathAnalyzer.enterNode (/Users/lachlan/code/projects/vue-test-utils/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:608:23)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
$
```

Hm.

Edit: Ok, this solved it: https://github.com/babel/babel-eslint/issues/681#issuecomment-420663038

**Comment by lmiller1990** [comment]
When I do `yarn test` locally is hangs on the Karma tests; is this the case for you locally @AtofStryker ? The karma process doesn't close. Is that normal?

**Comment by AtofStryker** [comment]
> When I do yarn test locally is hangs on the Karma tests; is this the case for you locally @AtofStryker ? The karma process doesn't close. Is that normal? 

@lmiller1990 Yeah that is expected. Karma tests by default run in watch mode, so when changes are made (locally) the tests rerun. If we don't want that behavior, we can just set `singleRun` to `true` [here](https://github.com/vuejs/vue-test-utils/pull/1629/files#diff-1f3a9e141f90ee6ca95a5229f4967f18R22)

> I am having many problems running this locally though. Linting failing...

Glad babel/babel-eslint#681 (comment) solved it. I didn't have issues linting locally, but I also reinstalled from lock after deduping. That also might help, but changing the rule should work just as well

I'll give the changes a pull locally and see if I have any issues

Edit: changes locally look good on my end with the latest commit.

**Comment by AtofStryker** [comment]
BTW @lmiller1990 were you having any other issues or was it just the linting?

**Comment by lmiller1990** [comment]
Everything else works great. I am guessing `singleRun` is somehow set to true on CI (or it would hang forever)?

It might be good to have it set locally by default (can you make that change?) I feel like other people contributing will have the same problem I just did.

Other than that, happy to merge and do a release!

**Comment by AtofStryker** [comment]
> Everything else works great. I am guessing `singleRun` is somehow set to true on CI (or it would hang forever)?

@lmiller1990 correct. The config is utilizing the circle var `process.env.CI`, which Circle sets to `true` when running in CI.

> It might be good to have it set locally by default (can you make that change?) I feel like other people contributing will have the same problem I just did.

Definitely! On revisiting, the watching behavior was probably a poor choice on my end :laughing: . Definitely don't want others to run into snags or problems when contributing. We can just set `singleRun: true` by default so we don't have to worry about any hanging. I have implemented in afad86e

> Other than that, happy to merge and do a release!

Sweet! We should be good to go then! :tada:  Once we get this merged in, I will open a few other PRs for some other library upgrades in, such as `vuepress`, `eslint`, `rollup`, etc. They are relatively minimal compared to this PR. Hopefully these changes allow for less cognitive overhead and context switching between the VTU 1.x and 2.x workspaces for contributors and the team :smiley:

**Comment by lmiller1990** [comment]
It's merged!!!! 🎉 thanks!

**Comment by lmiller1990** [comment]
It's out.

**Pull request merged**